lib/retry.go

```diff
@@ -0,0 +1,83 @@
+package lib
+
+import (
+	"time"
+)
+
+const (
+	defaultMinFailures = 0
+	defaultMaxWait     = 2 * time.Minute
+)
+
+// RetryWaiter will record failed and successful operations and provide
+// a channel to wait on before a failed operation can be retried.
+type RetryWaiter struct {
+	minFailures uint
+	maxWait     time.Duration
+	failures    uint
+}
+
+// Creates a new RetryWaiter
+func NewRetryWaiter(minFailures int, maxWait time.Duration) *RetryWaiter {
+	if minFailures < 0 {
+		minFailures = defaultMinFailures
+	}
+
+	if maxWait <= 0 {
+		maxWait = defaultMaxWait
+	}
+
+	return &RetryWaiter{
+		minFailures: uint(minFailures),
+		maxWait:     maxWait,
+		failures:    0,
+	}
+}
+
+// Marks that an operation is successful which resets the failure count.
+// The chan that is returned will be immediately selectable
+func (rw *RetryWaiter) Success() <-chan time.Time {
+	rw.Reset()
+	c := make(chan time.Time, 1)
+	c <- time.Now()
+	return c
+}
+
+// Marks that an operation failed. The chan returned will be selectable
+// once the calculated retry wait amount of time has elapsed
+func (rw *RetryWaiter) Failed() <-chan time.Time {
+	waitTime := 0 * time.Second
+	if rw.failures > rw.minFailures {
+		shift := rw.failures - rw.minFailures
+		waitTime = rw.maxWait
+		if shift < 31 {
+			waitTime = (1 << shift) * time.Second
+		}
+		if waitTime > rw.maxWait {
+			waitTime = rw.maxWait
+		}
+
+		// maybe add up to 10% extra
+		waitTime += RandomStagger(waitTime / 10)
+	}
+
+	// incrementing the failures after the waitTime calculation is
+	// intentional. This means that the first failure beyond the minimum
+	// failure threshold will cause a sleep of ~1s instead of ~2s.
+	rw.failures += 1
+
+	return time.After(waitTime)
+}
+
+func (rw *RetryWaiter) Reset() {
+	rw.failures = 0
+}
+
+// Wait is a convenience method to call either Success or Failed based
+// on a variable
+func (rw *RetryWaiter) Wait(failure bool) <-chan time.Time {
+	if failure {
+		return rw.Failed()
+	}
+	return rw.Success()
+}
```

Comment:
I know it's trivial but what would be nice to have at least a basic test case for this.

A couple of ideas that would make testing it easier (and I think benefit the lib generally):
 - make minWait configurable such that tests can use 10ms and still observe a few doublings without taking ages.
 - make jitter configurable so it can be disabled in tests (and generally more configurable for other use cases)

Don't think it's worth tons of time but seems like a pretty easy thing to add!

Reply:
The RetryWaiter now has tests.

I made the wait time calculation its own method so that it could be tested independently and have a handful of other tests that actually utilize the chans created.

I made an interface for things to provide their own Jitter calculations and have defined 1 type which adds a jitter based off of a percentage of the base wait time (same as it was doing before).